README.md
# Compilation on Ubuntu
In order to build the module correctly (tested on Ubuntu 16.04, kernel 4.4.0), you should have kernel source installed.
But since this is really a pain in the *ss and it's required only for building we will work around this.

You'll have to download the _linux-source_ package specific from your kernel from [security.ubuntu.com/ubuntu/pool/main/l/linux/](http://security.ubuntu.com/ubuntu/pool/main/l/linux/), extract it, navigate to its `usr/src/` folder, extract the archive and have it ready.
```
#!/bin/bash

# Get the filename
FILE=$(curl http://security.ubuntu.com/ubuntu/pool/main/l/linux/ 2> /dev/null | grep -Eo "linux\-source\-.*`uname -r | grep -Eo "[0-9\.\-]+" | sed "s/-$//"`" | grep -Eo '^[^"]+')

# Create a temp folder
FLD="/tmp/linux_source_`cat /dev/urandom | tr -dc 'a-zA-Z0-9' | fold -w 32 | head -n 1`"

# Download archive
wget -P $FLD "http://security.ubuntu.com/ubuntu/pool/main/l/linux/$FILE"

# Extract archive (to the Desktop for instance)
tar xjf $FLD/usr/src/linux-source-4.4.0.tar.bz2 -C ~/Desktop/

# Clean
rm -rf $FLD
```

At this point, building is easy. From your project git directory (see above for correct path)
```
export KSRC=~/Desktop/linux-source-4.4.0
make
```


Credits: https://askubuntu.com/a/510609


history.txt
A867 init from a333_evb1 on 2010/06/03

===================
Log
===================
name            yy/mm/dd

stylon		08/07/09	Initial import from eeepc-af903x-0.0.1

stylon		08/07/10	Convert from DOS to Unix CR/LF.

s001		08/07/10	1. Add support for A333 & A825.
				2. Fix compile warnings.
				3. A333 confirmed to work.
				4. No power management.

s002		08/07/10	Shorten channel change time by removing asking for
				lock after channel acquire.
				FILE: af903x-drv.c

s003		08/07/10	Implement BER, UBC and signal strength for debugging.
				FILE: af903x-drv.c af903x-fe.c af903x.h

stylon		08/07/11	Change Makefile to compile and package against EeePC
				2.6.21.4-eeepc kernel.
				FILE: Makefile

R00		08/07/11	Release 1.0.0 driver to ASUS.

j001		08/07/14	1. Modify the Makefile to prevent using the local DVB files.
				2. Remove the local DVB files.
				FILE: af903x.h Makefile insmod.sh rmmod.sh

s004		08/07/17	1. Implement monitor thread to recover from signal loss.
				2. Use monitor thread to read statistics.
				FILE: af903x.h af903x-fe.c af903x-drv.c 

s005		08/07/17	1. Supports selective suspend. 
				2. Tuner power control is also added.
				3. Fix possible race condition in power on/off.
				FILE: Philips_TDA18291HN.c af903x-core.c af903x.h 
				      af903x-devices.c af903x-fe.c af903x-drv.c

s006		08/07/17	1. Fix compile warnings.
				2. Add debug message to standard.c
				FILE: Common.h standard.c standard.h demodulatorextend.c
				      user.h mt2266.c type.h mt2266_userdef.h

s007		08/07/17	Remvoe Afatech string in USB driver.
				FILE: af903x-devices.h

s008		08/07/18	Update firmware from Windows driver.
				This seems enough to fix the issue resolved by s004.
				Undo monitor+relock in s004.
				Undo tuning before power down in s005.
				FILE: firmware.h

s009		08/07/18	Implement GPIO8 monitor algorithm.
				FILE: type.h af903x-drv.c

s010		08/07/18	Add IsPsbOverflow but not used.
				FILE: af903x-drv.c

s011		08/07/21	Supports 2.6.26 kernel.
				FILE: af903x-core.c

s012		08/07/21	1. Fix Makefile to remove Module.symvers in make clean
				2. Do not export symbols.
				FILE: af903x-fe.c af903x-tuner.c Makefile

stylon		08/07/21	Add AVerMedia versioning.

s013		08/07/29	Use semaphore to protect register access.
				FILE: af903x.h user.c

s014		08/07/29	Update firmware and tuner script from v8.8.27.0 EVB driver.
				FILE: Philips_TDA18291HN_Script.h firmware.h variable.h

s015		08/07/29	Restore monitor thread to reacquire channel when lock is lost.
				When signal is lost for quite a while, this is the only way
				to restore signal.
				FILE: af903x-drv.c af903x-fe.c

R01		08/08/07	Release v1.0.1 driver to ASUS.

s016		08/08/11	Add terminating entry as zero because it is
				not static variable.
				FILE: af903x-devices.c

s017		08/08/11	Increase dvb-usb buffer number to 10.
				FILE: af903x-devices.c
				
s018		08/08/11	Add proper error handling.
				FILE: af903x-fe.c

s019		08/08/11	Correctly report demod lock status via monitor thread.
				FILE: af903x-fe.c

s020		08/08/11	Report signal strength as EVB v0.0.3 driver
				does.
				FILE: af903x-fe.c

stylon		08/08/11	Update firmware and tuner script.
				FILE: firmware.h Philips_TDA18291HN_Script.h

s021		08/08/11	Implement DL_Reboot but do not use in af903x_suspend.
				FILE: af903x-drv.c 

s022		08/08/11	Use sleep instead of delay in User_delay.
				FILE: user.c

s023		08/08/11	Interpret bandwidth 6, 7, and 8 as 6M, 7M and 8M.
				This fixes some buggy AP not passing correct bandwidth value.
				FILE: af903x-fe.c

s024		08/08/11	Reject out-of-range frequency settings and consequetive setting
				frequency in +-125KHz.
				FILE: af903x-fe.c af903x-drv.c

s025		08/08/12	Change(relax) the criteria to reacquire channel because even if there is
				no signal there might be some random lock events.
				FILE: af903x-drv.c

s026		08/08/12	Use module paramter to decide if lock is always reported as true.
				FILE: af903x-fe.c 

s027		08/08/12	Enable debug message to allow all level of messages.
				Still must enable CONFIG_DVB_USB_DEBUG in debug.h first.
				FILE: af903x-drv.c

s028		08/08/13	Remove support for EVB. Can reenable using #define.
				FILE: af903x-devices.c af903x.h

s029		08/08/13	Undo part of s024 to not reject out-of-range frequency settings,
				FILE: af903x-fe.c

s030		08/08/14	Modify s029 so that we can report unlock after out-of-range frequency
				setting is rejected.
				FILE: af903x-fe.c

s031		08/08/22	Fix S3 resume on EeePC by issuing reboot to firmware.
				This only applies to kernel without support for selective suspend.
				FILE: af903x-fe.c af903x-core.c

j001		08/09/15	Detect the A337 product ID.
				FILE: af903x-core.c af903x-device.c

jamie		08/09/16	Add new tuner Maxlinear 5007 which is used by A337.
				FILE: Maxlinear_MXL5007.h Maxlinear_MXL5007.c Maxlinear_MXL5007_Script.h
					  MxL5007.h MxL5007.c MxL5007_API.h MxL5007_API.c MxL5007_Common.h

j002		08/09/17	Porting the windows code modification for A337.
				FILE: cmd.c demodulatorextend.h demodulatorextend.c standard.c af903x-drv.c

s032        	08/11/24    	Use 8.8.13.0 af903x firmware.
                        	FILE: firmware.h variable.h

j003		09/06/08	Support A867 (AF9035+MxL5007) hardware ID 0x0337
				FILE: af903x-core.c af903x-device.c af903x-drv.c

j004		09/06/09	Enable AF9035 HW PID filter.
				FILE: af903x-drv.c af903x-device.c af903x.h

j005		09/06/11	Add module parameter to enable/disable hw pid filter
				FILE: af903x-device.c

s033        	09/10/05    	Enable hw pid filter by default.
				FILE: af903x-device.c
                    

V1_0_4      	09/10/05	Release source code to PACE


j006		09/10/08	Workaround for HW pid filter: 
				Always add five PIDs: "0, 16, 17, 18, 20, 21" to HW pid table.
				FILE: af903x-device.c


V1_0_5      	09/10/08	Release source code to PACE


j007		09/10/23	Add SNR calculation. The value is not real SNR. It needs table look up.
				FILE: af903x-fe.c


V1_0_6		09/10/23	Release to Pace



Common.h
#pragma once

typedef unsigned char		BYTE;						/** At least 1 Byte				*/
typedef unsigned short		WORD;						/** At least 2 Bytes				*/
typedef unsigned int/*s006, long*/	DWORD;						/** At least 4 Bytes				*/
typedef void *				HANDLE;						/** Pointer to memory location	*/

#define TUNER_REGS_NUM		104
#define INITCTRL_NUM		38
#ifdef _MXL_PRODUCTION
#define CHCTRL_NUM			39
#else
#define CHCTRL_NUM			36
#endif

#define MXLCTRL_NUM			189

#define MASTER_CONTROL_ADDR		9

/** Enumeration of AGC Mode */
typedef enum
{
	MXL_DUAL_AGC = 0 ,
	MXL_SINGLE_AGC
} AGC_Mode ;

/**
 * Enumeration of Master Control Register State
 */
typedef enum
{
	MC_LOAD_START = 1 ,
	MC_POWER_DOWN ,
	MC_SYNTH_RESET ,
	MC_SEQ_OFF
} Master_Control_State ;

/**
 * Enumeration of MXL5005 Tuner Mode
 */
typedef enum
{
	MXL_ANALOG_MODE = 0 ,
	MXL_DIGITAL_MODE

} Tuner_Mode ;

/**
 * Enumeration of MXL5005 Tuner IF Mode
 */
typedef enum
{
	MXL_ZERO_IF = 0 ,
	MXL_LOW_IF

} Tuner_IF_Mode ;

/**
 * Enumeration of MXL5005 Tuner Clock Out Mode
 */
typedef enum
{
	MXL_CLOCK_OUT_DISABLE = 0 ,
	MXL_CLOCK_OUT_ENABLE
} Tuner_Clock_Out ;

/**
 * Enumeration of MXL5005 Tuner Div Out Mode
 */
typedef enum
{
	MXL_DIV_OUT_1 = 0 ,
	MXL_DIV_OUT_4

} Tuner_Div_Out ;

/**
 * Enumeration of MXL5005 Tuner Pull-up Cap Select Mode
 */
typedef enum
{
	MXL_CAP_SEL_DISABLE = 0 ,
	MXL_CAP_SEL_ENABLE

} Tuner_Cap_Select ;

/**
 * Enumeration of MXL5005 Tuner RSSI Mode
 */
typedef enum
{
	MXL_RSSI_DISABLE = 0 ,
	MXL_RSSI_ENABLE

} Tuner_RSSI ;

/**
 * Enumeration of MXL5005 Tuner Modulation Type
 */
typedef enum
{
	MXL_DEFAULT_MODULATION = 0 ,
	MXL_DVBT,
	MXL_ATSC,
	MXL_QAM

} Tuner_Modu_Type ;

/**
 * MXL5005 Tuner Register Struct
 */
typedef struct _TunerReg_struct
{
	WORD 	Reg_Num ;							/** Tuner Register Address */
	WORD	Reg_Val ;							/** Current sofware programmed value waiting to be writen */
} TunerReg_struct ;

/** 
 * MXL5005 Tuner Control Struct
 */
typedef struct _TunerControl_struct {
/*	char 	Name[40] ; */						/** Control Name */
	WORD	Ctrl_Num ;							/** Control Number */
	WORD	size ;								/** Number of bits to represent Value */
	WORD 	addr[25] ;							/** Array of Tuner Register Address for each bit position */
	WORD 	bit[25] ;							/** Array of bit position in Register Address for each bit position */
	WORD 	val[25] ;							/** Binary representation of Value */
} TunerControl_struct ;

/**
 * MXL5005 Tuner Struct
 */
typedef struct _Tuner_struct
{
	BYTE			Mode ;				/** 0: Analog Mode ; 1: Digital Mode */
	BYTE			IF_Mode ;			/** for Analog Mode, 0: zero IF; 1: low IF */
	DWORD			Chan_Bandwidth ;	/** filter  channel bandwidth (6, 7, 8) */
	DWORD			IF_OUT ;			/** Desired IF Out Frequency */
	WORD			IF_OUT_LOAD ;		/** IF Out Load Resistor (200/300 Ohms) */
	DWORD			RF_IN ;				/** RF Input Frequency */
	DWORD			Fxtal ;				/** XTAL Frequency */
	BYTE			AGC_Mode ;			/** AGC Mode 0: Dual AGC; 1: Single AGC */
	WORD			TOP ;				/** Value: take over point */
	BYTE			CLOCK_OUT ;			/** 0: turn off clock out; 1: turn on clock out */
	BYTE			DIV_OUT ;			/** 4MHz or 16MHz */
	BYTE			CAPSELECT ;			/** 0: disable On-Chip pulling cap; 1: enable */
	BYTE			EN_RSSI ;			/** 0: disable RSSI; 1: enable RSSI */
	BYTE			Mod_Type ;			/** Modulation Type; */
										/** 0 - Default;		1 - DVB-T */

	/** Calculated Settings */
	DWORD			RF_LO ;				/** Synth RF LO Frequency */
	DWORD			IF_LO ;				/** Synth IF LO Frequency */
	DWORD			TG_LO ;				/** Synth TG_LO Frequency */

	/** Pointers to ControlName Arrays */
	WORD					Init_Ctrl_Num ;					/** Number of INIT Control Names */
	TunerControl_struct		Init_Ctrl[INITCTRL_NUM] ;		/** INIT Control Names Array Pointer */
	WORD					CH_Ctrl_Num ;					/** Number of CH Control Names */
	TunerControl_struct		CH_Ctrl[CHCTRL_NUM] ;			/** CH Control Name Array Pointer */
	WORD					MXL_Ctrl_Num ;					/** Number of MXL Control Names */
	TunerControl_struct		MXL_Ctrl[MXLCTRL_NUM] ;			/** MXL Control Name Array Pointer */

	/** Pointer to Tuner Register Array */
	WORD					TunerRegs_Num ;		/** Number of Tuner Registers */
	TunerReg_struct			TunerRegs[TUNER_REGS_NUM] ;			/** Tuner Register Array Pointer */
} Tuner_struct ;



typedef enum
{
	/**
	 * Initialization Control Names
	 */
	DN_IQTN_AMP_CUT = 1 ,       /** 1  */
	BB_MODE ,                   /** 2  */
	BB_BUF ,                    /** 3  */
	BB_BUF_OA ,	                /** 4  */
	BB_ALPF_BANDSELECT ,        /** 5  */
	BB_IQSWAP ,                 /** 6  */
	BB_DLPF_BANDSEL ,           /** 7  */
	RFSYN_CHP_GAIN ,            /** 8  */
	RFSYN_EN_CHP_HIGAIN ,       /** 9  */
	AGC_IF ,                    /** 10 */
	AGC_RF ,                    /** 11 */
	IF_DIVVAL ,                 /** 12 */
	IF_VCO_BIAS ,               /** 13 */
	CHCAL_INT_MOD_IF ,          /** 14 */
	CHCAL_FRAC_MOD_IF ,         /** 15 */
	DRV_RES_SEL ,               /** 16 */
	I_DRIVER ,                  /** 17 */
	EN_AAF ,                    /** 18 */
	EN_3P ,                     /** 19 */
	EN_AUX_3P ,                 /** 20 */
	SEL_AAF_BAND ,              /** 21 */
	SEQ_ENCLK16_CLK_OUT ,       /** 22 */
	SEQ_SEL4_16B ,              /** 23 */
	XTAL_CAPSELECT ,            /** 24 */
	IF_SEL_DBL ,                /** 25 */
	RFSYN_R_DIV ,               /** 26 */
	SEQ_EXTSYNTHCALIF ,         /** 27 */
	SEQ_EXTDCCAL ,              /** 28 */
	AGC_EN_RSSI ,               /** 29 */
	RFA_ENCLKRFAGC ,            /** 30 */
	RFA_RSSI_REFH ,             /** 31 */
	RFA_RSSI_REF ,              /** 32 */
	RFA_RSSI_REFL ,             /** 33 */
	RFA_FLR ,                   /** 34 */
	RFA_CEIL ,                  /** 35 */
	SEQ_EXTIQFSMPULSE ,         /** 36 */
	OVERRIDE_1 ,                /** 37 */
	BB_INITSTATE_DLPF_TUNE,     /** 38 */
	/**
	 * Channel Change Control Names
	 */
	DN_POLY = 51 ,              /** 51 */
	DN_RFGAIN ,                 /** 52 */
	DN_CAP_RFLPF ,              /** 53 */
	DN_EN_VHFUHFBAR ,           /** 54 */
	DN_GAIN_ADJUST ,            /** 55 */
	DN_IQTNBUF_AMP ,            /** 56 */
	DN_IQTNGNBFBIAS_BST ,       /** 57 */
	RFSYN_EN_OUTMUX ,           /** 58 */
	RFSYN_SEL_VCO_OUT ,         /** 59 */
	RFSYN_SEL_VCO_HI ,          /** 60 */
	RFSYN_SEL_DIVM ,            /** 61 */
	RFSYN_RF_DIV_BIAS ,         /** 62 */
	DN_SEL_FREQ ,               /** 63 */
	RFSYN_VCO_BIAS ,            /** 64 */
	CHCAL_INT_MOD_RF ,          /** 65 */
	CHCAL_FRAC_MOD_RF ,         /** 66 */
	RFSYN_LPF_R ,               /** 67 */
	CHCAL_EN_INT_RF ,           /** 68 */
	TG_LO_DIVVAL ,              /** 69 */
	TG_LO_SELVAL ,              /** 70 */
	TG_DIV_VAL ,                /** 71 */
	TG_VCO_BIAS ,               /** 72 */
	SEQ_EXTPOWERUP ,            /** 73 */
	OVERRIDE_2 ,                /** 74 */
	OVERRIDE_3 ,                /** 75 */
	OVERRIDE_4 ,                /** 76 */
	SEQ_FSM_PULSE ,             /** 77 */
	GPIO_4B,                    /** 78 */
	GPIO_3B,                    /** 79 */
	GPIO_4,                     /** 80 */
	GPIO_3,                     /** 81 */
	GPIO_1B,                    /** 82 */
	DAC_A_ENABLE ,              /** 83 */
	DAC_B_ENABLE ,              /** 84 */
	DAC_DIN_A ,	                /** 85 */
	DAC_DIN_B ,                 /** 86 */
#ifdef _MXL_PRODUCTION
	RFSYN_EN_DIV,				/** 87 */
	RFSYN_DIVM,					/** 88 */
	DN_BYPASS_AGC_I2C			/** 89 */
#endif

} MXL5005_ControlName ;
